pkg/controller/core.oam.dev/v1alpha2/application/parser/service.go

```diff
@@ -0,0 +1,322 @@
+	WithTemplate(raw string) Render
+	Complete() (*cue.Instance, error)
+}
+
+// Workload is component
+type Workload struct {
+	name     string
+	typ      string
+	params   map[string]interface{}
+	template string
+	traits   []*Trait
+}
+
+// Name export workload's name
+func (wl *Workload) Name() string {
+	return wl.name
+}
+
+// Traits export workload's traits
+func (wl *Workload) Traits() []*Trait {
+	return wl.traits
+}
+
+// Eval convert template to Component
+func (wl *Workload) Eval(render Render) (*v1alpha2.Component, error) {
+	inst, err := render.WithParams(wl.params).WithTemplate(wl.template).Complete()
+	if err != nil {
+		return nil, errors.WithMessagef(err, "component %s eval", wl.name)
+	}
+	output := inst.Lookup("output")
+	if !output.Exists() {
+		return nil, errors.Errorf("output not found in component %s ", wl.name)
+	}
+
+	data, err := cueJson.Marshal(output)
+	if err != nil {
+		return nil, errors.WithMessagef(err, "component %s marshal", wl.name)
+	}
+
+	componentRef := &unstructured.Unstructured{}
+	if err := componentRef.UnmarshalJSON([]byte(data)); err != nil {
+		return nil, errors.WithMessagef(err, "component %s UnmarshalJSON to unstructured", wl.name)
+	}
+
+	component := new(v1alpha2.Component)
+
+	component.Spec.Workload.Object = componentRef
+	return component, nil
+}
+
+// Trait is ComponentTrait
+type Trait struct {
+	name     string
+	params   map[string]interface{}
+	template string
+}
+
+// Eval convert template to ComponentTrait
+func (trait *Trait) Eval(render Render) ([]v1alpha2.ComponentTrait, error) {
+	inst, err := render.WithParams(trait.params).WithTemplate(trait.template).Complete()
+	if err != nil {
+		return nil, errors.WithMessagef(err, "traitDef %s eval", trait.name)
+	}
+	cueValues := []cue.Value{}
+	output := inst.Lookup("output")
+	if !output.Exists() {
+
+		outputs := inst.Lookup("outputs")
+		iter, err := outputs.List()
+		if err != nil {
+			return nil, errors.Errorf("output|outputs not found in traitDef %s ", trait.name)
+			// return nil,errors.WithMessagef(err,"traitDef %s outputs must be list",trait.name)
+		}
+		for iter.Next() {
+			cueValues = append(cueValues, iter.Value())
+		}
+
+	} else {
+		cueValues = append(cueValues, output)
+	}
+
+	if len(cueValues) == 0 {
+		return nil, errors.Errorf("output|outputs not found in traitDef %s ", trait.name)
+	}
+
+	compTraits := []v1alpha2.ComponentTrait{}
+	for _, cv := range cueValues {
+		data, err := cueJson.Marshal(cv)
+		if err != nil {
+			return nil, errors.WithMessagef(err, "traitDef %s marshal", trait.name)
+		}
+		traitRef := &unstructured.Unstructured{}
+		if err := traitRef.UnmarshalJSON([]byte(data)); err != nil {
+			return nil, errors.WithMessagef(err, "traitDef %s UnmarshalJSON to unstructured", trait.name)
+		}
+		compTrait := v1alpha2.ComponentTrait{}
+		compTrait.Trait.Object = traitRef
+		compTraits = append(compTraits, compTrait)
+	}
+
+	return compTraits, nil
+}
+
+// Appfile describle application
+type Appfile struct {
```

Comment:
Why don't use the existing AppFile struct?

Reply:
The current appfile structure is an open structure of map does not meet the coding requirements